<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bamnostic package &mdash; bamnostic 1.1.8
 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Citation and Support" href="support.html" />
    <link rel="prev" title="Quickstart" href="quickstart.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> bamnostic
          </a>
              <div class="version">
                1.1.8

              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Module documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.core">Core Module (<code class="docutils literal notranslate"><span class="pre">bamnostic.core</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.bgzf">Bgzip BAM Format (BGZF) file handlers (<code class="docutils literal notranslate"><span class="pre">bamnostic.bgzf</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.bam">Binary Alignment Map (BAM) file handlers (<code class="docutils literal notranslate"><span class="pre">bamnostic.bam</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.bai">BAM Index (BAI) file handler (<code class="docutils literal notranslate"><span class="pre">bamnostic.bai</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.csi">Coordinate-Sorted Index (CSI) file handler (<code class="docutils literal notranslate"><span class="pre">bamnostic.csi</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-bamnostic.utils">BAMnostic Utilities (<code class="docutils literal notranslate"><span class="pre">bamnostic.utils</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="support.html">Citation and Support</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">bamnostic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>bamnostic package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/bamnostic.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="module-bamnostic">
<span id="bamnostic-package"></span><h1>bamnostic package<a class="headerlink" href="#module-bamnostic" title="Permalink to this headline">Â¶</a></h1>
<p><cite>bamnostic</cite> is an OS-agnostic, Pure Python <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a>
file parser.</p>
<p>The purpose of <cite>bamnostic</cite> is to open up BAM query capabilities to all current
OS environments. Special care was taken to allow <cite>bamnostic</cite> to run in all versions
of Python 2.7 onward. Furthermore, <cite>bamnostic</cite> does not require any packages
outside of the standard library. Therefore, <cite>bamnostic</cite> will run on all stable
versions of <a class="reference external" href="https://pypy.org/">PyPy</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>SAM and CRAM support is not yet implemented</p>
</div>
<p>The three main classes of <cite>bamnostic</cite> are:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bamnostic.AlignmentFile</span></code>: the BAM file handler</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bamnostic.AlignedSegment</span></code>: an aligned read object interface</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">bamnostic.bai.Bai</span></code>: if the BAM file has an associated index file (preferred),     this is the file handler for it.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Within the scope of personal research, reading BAM files is the only fully
supported IO. The skeleton for writing BAM files is present, just not connected.</p>
</div>
<p>This code is part of the bamnostic distribution and governed by its
license.  Please see the LICENSE file that should have been included
as part of this package.</p>
<div class="section" id="module-bamnostic.core">
<span id="core-module-bamnostic-core"></span><h2>Core Module (<code class="docutils literal notranslate"><span class="pre">bamnostic.core</span></code>)<a class="headerlink" href="#module-bamnostic.core" title="Permalink to this headline">Â¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.core.Cigar">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">Cigar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">op_code</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_op</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">op_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">op_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.core.Cigar" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">namedtuple</span></code> for handling CIGAR data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>op_code</strong> (<em>int</em>) â€“ CIGAR operation index</p></li>
<li><p><strong>n_op</strong> (<em>int</em>) â€“ number of operations for a given op_code</p></li>
<li><p><strong>op_id</strong> (<em>str</em>) â€“ the string representation of the CIGAR representation (â€˜MIDNSHP=XBâ€™)</p></li>
<li><p><strong>op_name</strong> (<em>str</em>) â€“ Longer string name for operation</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.Cigar.n_op">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">n_op</span></span><a class="headerlink" href="#bamnostic.core.Cigar.n_op" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.Cigar.op_code">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">op_code</span></span><a class="headerlink" href="#bamnostic.core.Cigar.op_code" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.Cigar.op_id">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">op_id</span></span><a class="headerlink" href="#bamnostic.core.Cigar.op_id" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.Cigar.op_name">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">op_name</span></span><a class="headerlink" href="#bamnostic.core.Cigar.op_name" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.core.offset_qual">
<span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">offset_qual</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">qual_string</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#offset_qual"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.offset_qual" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Offsets the ASCII-encoded quality string to represent PHRED score.</p>
<p>Every base that is in the alignment is assigned a Phred score. A Phred
score (<span class="math notranslate nohighlight">\(Q\)</span>) is defined as <span class="math notranslate nohighlight">\(Q=-10\log_{10}P\)</span>, where <span class="math notranslate nohighlight">\(P\)</span>
is the base-calling error probability. Phred quality scores tend range
from 10 to 60. These qualities are then offset by 33 and ASCII-encoded
for readability and storage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>qual_string</strong> (<code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code> or <code class="xref py py-obj docutils literal notranslate"><span class="pre">bytes</span></code>) â€“ Phred quality scores without offset</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>ASCII-encoded Phred scores offest by adding 33 to base score.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(str)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">qual_score</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">offset_qual</span><span class="p">(</span><span class="n">qual_score</span><span class="p">))</span>
<span class="go">&#39;&lt;&lt;&lt;7&lt;&lt;&lt;;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;8;;&lt;7;4&lt;;&lt;;;;;;94&lt;;&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.core.AlignmentFile">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">AlignmentFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_or_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">128</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_sq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">duplicate_filehandle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_truncation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compresslevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">b''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignmentFile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignmentFile" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#bamnostic.bam.BamReader" title="bamnostic.bam.BamReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.bam.BamReader</span></code></a>, <a class="reference internal" href="#bamnostic.bam.BamWriter" title="bamnostic.bam.BamWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.bam.BamWriter</span></code></a></p>
<p>Wrapper to allow drop in replacement for BAM functionality in a <code class="docutils literal notranslate"><span class="pre">pysam</span></code>-like API.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath_or_object</strong> (str | <code class="xref py py-obj docutils literal notranslate"><span class="pre">file</span></code>) â€“ the path or file object of the BAM file</p></li>
<li><p><strong>mode</strong> (<em>str</em>) â€“ Mode for reading. BAM files are binary by nature (default: â€˜rbâ€™).</p></li>
<li><p><strong>max_cache</strong> (<em>int</em>) â€“ number of desired LRU cache size, preferably a multiple of 2 (default: 128).</p></li>
<li><p><strong>index_filename</strong> (<em>str</em>) â€“ path to index file (BAI) if it is named differently than the BAM file (default: None).</p></li>
<li><p><strong>filename</strong> (str | <code class="xref py py-obj docutils literal notranslate"><span class="pre">file</span></code>) â€“ synonym for <cite>filepath_or_object</cite></p></li>
<li><p><strong>check_header</strong> (<em>bool</em>) â€“ Obsolete method maintained for backwards compatibility (default: False)</p></li>
<li><p><strong>check_sq</strong> (<em>bool</em>) â€“ Inspect BAM file for <cite>&#64;SQ</cite> entries within the header</p></li>
<li><p><strong>reference_filename</strong> (<em>str</em>) â€“ Not implemented. Maintained for backwards compatibility</p></li>
<li><p><strong>filepath_index</strong> (<em>str</em>) â€“ synonym for <cite>index_filename</cite></p></li>
<li><p><strong>require_index</strong> (<em>bool</em>) â€“ require the presence of an index file or raise (default: False)</p></li>
<li><p><strong>duplicate_filehandle</strong> (<em>bool</em>) â€“ Not implemented. Raises warning if True.</p></li>
<li><p><strong>ignore_truncation</strong> (<em>bool</em>) â€“ Whether or not to allow trucated file processing (default: False).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">AlignedSegment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_io</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Main class for handling reads within the BAM</p>
<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_alignment_end">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_alignment_end</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_alignment_end" title="Permalink to this definition">Â¶</a></dt>
<dd><p>One past the final index position of the aligned portion of the read</p>
<p><cite>query_alignment_*</cite> all refer to the portion of the read that was aligned,
and therefore exclude clipping, but include insertions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_alignment_start">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_alignment_start</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_alignment_start" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The starting index of the aligned portion of the read</p>
<p><cite>query_alignment_*</cite> all refer to the portion of the read that was aligned,
and therefore exclude clipping, but include insertions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_name">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_name</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_name" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for the read name</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_alignment_sequence">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_alignment_sequence</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_alignment_sequence" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The sequence of the aligned portion of the read</p>
<p><cite>query_alignment_*</cite> all refer to the portion of the read that was aligned,
and therefore exclude clipping, but include insertions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_alignment_length">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_alignment_length</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_alignment_length" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The length of the aligned portion of the read</p>
<p><cite>query_alignment_*</cite> all refer to the portion of the read that was aligned,
and therefore exclude clipping, but include insertions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.reference_start">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">reference_start</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.reference_start" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The starting index of the aligned portion of the read</p>
<p><cite>reference_</cite> here means the portion of the read that was aligned
according to the reference. Therefore, does not include insertions
and clipping</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.reference_end">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">reference_end</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.reference_end" title="Permalink to this definition">Â¶</a></dt>
<dd><p>One past the final index position of the aligned portion of the read</p>
<p><cite>reference_</cite> here means the portion of the read that was aligned
according to the reference. Therefore, does not include insertions
and clipping</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.reference_length">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">reference_length</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.reference_length" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The length of the aligned portion of the read</p>
<p><cite>reference_</cite> here means the portion of the read that was aligned
according to the reference. Therefore, does not include insertions
and clipping</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_sequence">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_sequence</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_sequence" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The sequence of the read</p>
<p><cite>query_</cite> here means the query as seen in the BAM file, and therefore
includes clipping</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.query_length">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">query_length</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.query_length" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The length of sequence of the read</p>
<p><cite>query_</cite> here means the query as seen in the BAM file, and therefore
includes clipping</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.next_reference_id">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">next_reference_id</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.next_reference_id" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the reference id of mate read</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.next_reference_name">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">next_reference_name</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.next_reference_name" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the reference name of mate read</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.next_reference_start">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">next_reference_start</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.next_reference_start" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the reference name of mate read</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_duplicate">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_duplicate</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_duplicate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as PCR or optical duplicate according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_paired">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_paired</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_paired" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as paired according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_proper_pair">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_proper_pair</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_proper_pair" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If each segment properly aligned according to the aligner</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_qcfail">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_qcfail</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_qcfail" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as QC fail according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_read1">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_read1</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_read1" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as the first read of mate pair according to read flag (implies the read is mated)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_read2">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_read2</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_read2" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as the second read of mate pair according to read flag (implies the read is mated)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_reverse">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_reverse</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as reversed according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_secondary">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_secondary</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_secondary" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as secondary alignment according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_supplementary">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_supplementary</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_supplementary" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as a supplementary alignment according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.is_unmapped">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">is_unmapped</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.is_unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>If the read is set as unmapped according to read flag</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.mate_is_reverse">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">mate_is_reverse</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.mate_is_reverse" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Check the read flag to see if the mat is reverse strand (implies the read is mated)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.mate_is_unmapped">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">mate_is_unmapped</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.mate_is_unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Check the read flag to see if mate is unmapped (implies the read is mated)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.mapping_quality">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">mapping_quality</span></span><a class="headerlink" href="#bamnostic.core.AlignedSegment.mapping_quality" title="Permalink to this definition">Â¶</a></dt>
<dd><p>The mapping quality (MAPQ) of the read</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.get_tag">
<span class="sig-name descname"><span class="pre">get_tag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tag</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_value_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment.get_tag"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment.get_tag" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Gets the value associated with a given tag key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag</strong> (<em>str</em>) â€“ the tag of interest</p></li>
<li><p><strong>with_value_type</strong> (<em>bool</em>) â€“ return what kind of value the tag</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the value associated with a given tag or the value and type
of value (as seen in BAM format)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.get_tags">
<span class="sig-name descname"><span class="pre">get_tags</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">with_value_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment.get_tags"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment.get_tags" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Returns all the tags for a given read</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>with_value_type</strong> (<em>bool</em>) â€“ return the tag value type (as defined by BAM format)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tag tuples (with or without tag value type)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>f_tags(<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.get_cigar_stats">
<span class="sig-name descname"><span class="pre">get_cigar_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment.get_cigar_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment.get_cigar_stats" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Gets the counts of each CIGAR operation in the read and number of
nucleotides related to those given operations.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of CIGAR operation counts
nt_counts (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code>): list of nucleotide counts for each operation</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>op_blocks (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.ref_gen">
<span class="sig-name descname"><span class="pre">ref_gen</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment.ref_gen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment.ref_gen" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Recreates the reference sequence associated with the given segment.</p>
<p>Uses the CIGAR string and MD tag to recreate the reference sequence associated
with the aligned segment. This is done without the need for looking up
the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generated reference sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(str)</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>KeyError</strong> â€“ if read does not contain MD tag</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.core.AlignedSegment.to_bam">
<span class="sig-name descname"><span class="pre">to_bam</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/core.html#AlignedSegment.to_bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.core.AlignedSegment.to_bam" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Writes the alignment record to a BAM file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bam_file</strong> (string or <a class="reference internal" href="#bamnostic.bam.BamWriter" title="bamnostic.bam.BamWriter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bamnostic.bam.BamWriter</span></code></a>) â€“ BAM file path or open bam file in a write mode</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.core.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.core.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.core.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.core.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.core.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

</div>
<div class="section" id="module-bamnostic.bgzf">
<span id="bgzip-bam-format-bgzf-file-handlers-bamnostic-bgzf"></span><h2>Bgzip BAM Format (BGZF) file handlers (<code class="docutils literal notranslate"><span class="pre">bamnostic.bgzf</span></code>)<a class="headerlink" href="#module-bamnostic.bgzf" title="Permalink to this headline">Â¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_gzip_header">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_gzip_header</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_gzip_header" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_subfields">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_subfields</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_subfields" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_gzip_integrity">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_gzip_integrity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_gzip_integrity" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_bgzf_metaheader">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_bgzf_metaheader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_bgzf_metaheader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.get_block">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">get_block</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#get_block"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.get_block" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Pulls out entire GZIP block</p>
<p>Used primarily for copying the header block of a BAM file. However,
it can be used to copy any BGZF block within a BAM file that starts at
the given offset.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Does not progress file cursor position.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>handle</strong> (<code class="xref py py-obj docutils literal notranslate"><span class="pre">file</span></code>) â€“ open BAM file</p></li>
<li><p><strong>offset</strong> (<em>int</em>) â€“ offset of BGZF block (default: 0)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Complete BGZF block</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError</strong> â€“ if the BGZF block header is malformed</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam_header</span> <span class="o">=</span> <span class="n">get_block</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">try</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">bam_header</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="sa">b</span><span class="s1">&#39;</span><span class="se">\x1f\x8b\x08\x04</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="gp">... </span><span class="k">except</span> <span class="ne">SyntaxError</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">bam_header</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\x1f\x8b\x08\x04</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">BgzfReader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_or_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">duplicate_filehandle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_truncation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>The BAM reader. Heavily modified from Peter Cockâ€™s BgzfReader.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.tell">
<span class="sig-name descname"><span class="pre">tell</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.tell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.tell" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a 64-bit unsigned BGZF virtual offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.seek">
<span class="sig-name descname"><span class="pre">seek</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">virtual_offset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.seek"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.seek" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Seek to a 64-bit unsigned BGZF virtual offset.</p>
<p>A virtual offset is a composite number made up of the compressed
offset (<cite>coffset</cite>) position of the start position of the BGZF block that
the position originates within, and the uncompressed offset (<cite>uoffset</cite>)
within the deflated BGZF block where the position starts. The virtual offset
is defined as</p>
<p><cite>virtual_offset = coffset &lt;&lt; 16 | uoffset</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>virtual_offset</strong> (<em>int</em>) â€“ 64-bit unsigned composite byte offset</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>an echo of the new position</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>virtual_offset (int)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> â€“ if within block offset is more than block size</p></li>
<li><p><strong>AssertionError</strong> â€“ if the start position is not the block start position</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">10</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">42</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Within offset 42 but block size only 38</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.read" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Read method for the BGZF module.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>size</strong> (<em>int</em>) â€“ the number of bytes to read from file. Advances the cursor.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>byte string of length <cite>size</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>data (<code class="xref py py-obj docutils literal notranslate"><span class="pre">bytes</span></code>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>NotImplementedError</strong> â€“ if the user tries to read the whole file</p></li>
<li><p><strong>AssertionError</strong> â€“ if read does not return any data</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.readline">
<span class="sig-name descname"><span class="pre">readline</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.readline"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.readline" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Read a single line for the BGZF file.</p>
<p>Binary operations do not support <cite>readline()</cite>. Code is commented
out for posterity sake</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.close" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Close BGZF file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.seekable">
<span class="sig-name descname"><span class="pre">seekable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.seekable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.seekable" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True indicating the BGZF supports random access.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Modified from original Bio.BgzfReader: checks to see if BAM
file has associated index file (BAI)</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.isatty">
<span class="sig-name descname"><span class="pre">isatty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.isatty"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.isatty" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if connected to a TTY device.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfReader.fileno">
<span class="sig-name descname"><span class="pre">fileno</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfReader.fileno"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfReader.fileno" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return integer file descriptor.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">BgzfWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_or_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'wb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compresslevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Define a BGZFWriter object.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.write" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write method for the class.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.flush"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.flush" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Flush data explicitly.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.close" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Flush data, write 28 bytes BGZF EOF marker, and close BGZF file.</p>
<p>samtools will look for a magic EOF marker, just a 28 byte empty BGZF
block, and if it is missing warns the BAM file may be truncated. In
addition to samtools writing this block, so too does bgzip - so this
implementation does too.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.tell">
<span class="sig-name descname"><span class="pre">tell</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.tell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.tell" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a BGZF 64-bit virtual offset.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.seekable">
<span class="sig-name descname"><span class="pre">seekable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.seekable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.seekable" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True indicating the BGZF supports random access.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.isatty">
<span class="sig-name descname"><span class="pre">isatty</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.isatty"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.isatty" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True if connected to a TTY device.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bgzf.BgzfWriter.fileno">
<span class="sig-name descname"><span class="pre">fileno</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bgzf.html#BgzfWriter.fileno"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bgzf.BgzfWriter.fileno" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return integer file descriptor.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bgzf.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bgzf.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bgzf.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

</div>
<div class="section" id="module-bamnostic.bam">
<span id="binary-alignment-map-bam-file-handlers-bamnostic-bam"></span><h2>Binary Alignment Map (BAM) file handlers (<code class="docutils literal notranslate"><span class="pre">bamnostic.bam</span></code>)<a class="headerlink" href="#module-bamnostic.bam" title="Permalink to this headline">Â¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bam.</span></span><span class="sig-name descname"><span class="pre">BAMheader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_io</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BAMheader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BAMheader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Parse and store the BAM file header</p>
<p>The BAM header is the plain text and byte-encoded metadata of a given BAM file.
Information stored in the header are the number, length, and name of the reference
sequences that reads were aligned to; version of software used; read group identifiers; etc.
The <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a> format also stipulates that the first block of any BAM file should be reserved
just for the BAM header block.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader._io">
<span class="sig-name descname"><span class="pre">_io</span></span><a class="headerlink" href="#bamnostic.bam.BAMheader._io" title="Permalink to this definition">Â¶</a></dt>
<dd><p>opened BAM file object</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">file</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader.SAMheader">
<span class="sig-name descname"><span class="pre">SAMheader</span></span><a class="headerlink" href="#bamnostic.bam.BAMheader.SAMheader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>parsed dictionary of the SAM header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">dict</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader.n_refs">
<span class="sig-name descname"><span class="pre">n_refs</span></span><a class="headerlink" href="#bamnostic.bam.BAMheader.n_refs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of references</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader.refs">
<span class="sig-name descname"><span class="pre">refs</span></span><a class="headerlink" href="#bamnostic.bam.BAMheader.refs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>reference names and lengths listed in the BAM header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">dict</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BAMheader.to_header">
<span class="sig-name descname"><span class="pre">to_header</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BAMheader.to_header"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BAMheader.to_header" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Allows the user to directly copy the header of another BAM file</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>packed byte code of entire header BGZF block</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(bytesarray)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bam.</span></span><span class="sig-name descname"><span class="pre">BamReader</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_or_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_cache</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">128</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_sq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">require_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">duplicate_filehandle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_truncation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#bamnostic.bgzf.BgzfReader" title="bamnostic.bgzf.BgzfReader"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.bgzf.BgzfReader</span></code></a></p>
<p>The BAM reader. Heavily modified from Peter Cockâ€™s BgzfReader.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.header">
<span class="sig-name descname"><span class="pre">header</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.header" title="Permalink to this definition">Â¶</a></dt>
<dd><p>representation of header data (if present)</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.lengths">
<span class="sig-name descname"><span class="pre">lengths</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.lengths" title="Permalink to this definition">Â¶</a></dt>
<dd><p>lengths of references listed in header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.nocoordinate">
<span class="sig-name descname"><span class="pre">nocoordinate</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.nocoordinate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of reads that have no coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.nreferences">
<span class="sig-name descname"><span class="pre">nreferences</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.nreferences" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of references in header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.ref2tid">
<span class="sig-name descname"><span class="pre">ref2tid</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.ref2tid" title="Permalink to this definition">Â¶</a></dt>
<dd><p>refernce names and refID dictionary</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">dict</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">int</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.references">
<span class="sig-name descname"><span class="pre">references</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.references" title="Permalink to this definition">Â¶</a></dt>
<dd><p>names of references listed in header</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.text">
<span class="sig-name descname"><span class="pre">text</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.text" title="Permalink to this definition">Â¶</a></dt>
<dd><p>SAM header (if present)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.unmapped">
<span class="sig-name descname"><span class="pre">unmapped</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of unmapped reads</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This implementation is likely to change. While the API was meant to
mirror <cite>pysam</cite>, it makes sense to include the <cite>pysam</cite>-like API in an extension
that will wrap the core reader. This would be a major refactor, and therefore
will not happen any time soon (30 May 2018).</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.check_index">
<span class="sig-name descname"><span class="pre">check_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index_filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">req_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.check_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.check_index" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks to make sure index file is available. If not, it disables random access.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index_filename</strong> (<em>str</em>) â€“ path to index file (BAI) if it does not fit naming convention (default: None).</p></li>
<li><p><strong>req_idx</strong> (<em>bool</em>) â€“ Raise error if index file is not present (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if index is present, else False</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(bool)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>IOError</strong> â€“ If the index file is closed or index could not be opened</p>
</dd>
<dt class="field-odd">Warns</dt>
<dd class="field-odd"><p><strong>UserWarning</strong> â€“ If index could not be loaded. Random access is disabled.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">check_index</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span> <span class="o">+</span> <span class="s1">&#39;.bai&#39;</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">check_index</span><span class="p">(</span><span class="s1">&#39;not_a_file.bai&#39;</span><span class="p">)</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id0">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">nocoordinate</span></span><a class="headerlink" href="#id0" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get the number of reads without coordiantes according to the statistics recorded in the index.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>sum of reads without coordinates</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.mapped">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">mapped</span></span><a class="headerlink" href="#bamnostic.bam.BamReader.mapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get the number of mapped reads according  to the statistics recorded in the index.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>sum of mapped reads</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id2">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">unmapped</span></span><a class="headerlink" href="#id2" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get the number of unmapped reads without coordiantes <strong>and</strong>
without coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>sum of unmapped reads and reads without coordinates</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id3">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">references</span></span><a class="headerlink" href="#id3" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get all references used in alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>reference names</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">tuple</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id4">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">nreferences</span></span><a class="headerlink" href="#id4" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get the number of references listed in alignment</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>count of references</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="id5">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">lengths</span></span><a class="headerlink" href="#id5" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get all reference lengths used in alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>reference lengths</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">tuple</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">int</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.has_index">
<span class="sig-name descname"><span class="pre">has_index</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.has_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.has_index" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks if file has index and it is open</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if present and opened, else False</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.fetch">
<span class="sig-name descname"><span class="pre">fetch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">until_eof</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple_iterators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.fetch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.fetch" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Creates a generator that returns all reads within the given region. (inclusive, exclusive)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) â€“ name of reference/contig</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ start position of region of interest (0-based)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ stop position of region of interest (0-based)</p></li>
<li><p><strong>region</strong> (<em>str</em>) â€“ SAM region formatted string. Accepts tab-delimited values as well</p></li>
<li><p><strong>tid</strong> (<em>int</em>) â€“ the refID or target id of a reference/contig</p></li>
<li><p><strong>until_eof</strong> (<em>bool</em>) â€“ iterate until end of file</p></li>
<li><p><strong>mutiple_iterators</strong> (<em>bool</em>) â€“ allow multiple iterators over region. Not Implemented.                  Notice: each iterator will open up a new view into the BAM file, so overhead will apply.</p></li>
<li><p><strong>reference</strong> (<em>str</em>) â€“ synonym for <cite>contig</cite></p></li>
<li><p><strong>end</strong> (<em>str</em>) â€“ synonym for <cite>stop</cite></p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>reads over the region of interest if any</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError</strong> â€“ if the genomic coordinates are out of range or invalid</p></li>
<li><p><strong>KeyError</strong> â€“ Reference is not found in header</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>SAM region formatted strings take on the following form:
â€˜chr1:100000-200000â€™</p>
</div>
<dl>
<dt>Usage:</dt><dd><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">AlignmentFile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">AlignmentFile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">AlignmentFile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1:1-1000&#39;</span><span class="p">)</span>
<span class="n">AlignmentFile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">AlignmentFile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span> 
<span class="go">EAS56_57:6:190:289:82 ... MF:C:192</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;chr10 was not found in the file header&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1700</span><span class="p">,</span> <span class="mi">1701</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Genomic region out of bounds.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">AssertionError</span>: <span class="n">Malformed region: start should be &lt;= stop, you entered 100, 10</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">until_eof</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nofilter'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.count" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Count the number of reads in the given region</p>
<p>Note: this counts the number of reads that <strong>overlap</strong> the given region.</p>
<p>Can potentially make use of a filter for the reads (or custom function
that returns <cite>True</cite> or <cite>False</cite> for each read).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) â€“ the reference name (Default: None)</p></li>
<li><p><strong>reference</strong> (<em>str</em>) â€“ synonym for <cite>contig</cite> (Default: None)</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ 0-based inclusive start position (Default: None)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ 0-based exclusive start position (Default: None)</p></li>
<li><p><strong>end</strong> (<em>int</em>) â€“ Synonym for <cite>stop</cite> (Default: None)</p></li>
<li><p><strong>region</strong> (<em>str</em>) â€“ SAM-style region format. Example: â€˜chr1:10000-50000â€™ (Default: None)</p></li>
<li><p><strong>until_eof</strong> (<em>bool</em>) â€“ count number of reads from start to end of file Note, this can potentially be an expensive operation. (Default: False)</p></li>
<li><p><strong>read_callback</strong> (<em>str</em><em>|</em><em>function</em>) â€“ <p>select (or create) a filter of which reads to count. Built-in filters:</p>
<ul>
<li><p><cite>all</cite>: skips reads that contain the following flags:</p>
<blockquote>
<div><ul class="simple">
<li><p>0x4 (4): read unmapped</p></li>
<li><p>0x100 (256): not primary alignment</p></li>
<li><p>0x200 (512): QC Fail</p></li>
<li><p>0x400 (1024): PCR or optical duplicate</p></li>
</ul>
</div></blockquote>
</li>
<li><p><cite>nofilter</cite>: uses all reads (Default)</p></li>
<li><dl class="simple">
<dt>The user can also supply a custom function that </dt><dd><p>returns boolean objects for each read</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>count of reads in the given region that meet parameters</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> â€“ if genomic coordinates are out of range or invalid or random access is disabled</p></li>
<li><p><strong>RuntimeError</strong> â€“ if <cite>read_callback</cite> is not properly set</p></li>
<li><p><strong>KeyError</strong> â€“ Reference is not found in header</p></li>
<li><p><strong>AssertionError</strong> â€“ if genomic region is malformed</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>SAM region formatted strings take on the following form:
â€˜chr1:100000-200000â€™</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>AlignmentFile.count(contig=â€™chr1â€™, start=1, stop= 1000)
AlignmentFile.count(â€˜chr1â€™, 1, 1000)
AlignmentFile.count(â€˜chr1:1-1000â€™)
AlignmentFile.count(â€˜chr1â€™, 1)
AlignmentFile.count(â€˜chr1â€™)</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="n">read_callback</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;chr10 was not found in the file header&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">1700</span><span class="p">,</span> <span class="mi">1701</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Genomic region out of bounds.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">AssertionError</span>: <span class="n">Malformed region: start should be &lt;= stop, you entered 100, 10</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.count_coverage">
<span class="sig-name descname"><span class="pre">count_coverage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_quality_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.count_coverage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.count_coverage" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Counts the coverage of each base supported by a read the given interval.</p>
<p>Given an interval (inclusive, exclusive), this method pulls each read that overlaps
with the region. To ensure that the read truly overlaps with the region, the CIGAR string
is required. These reads can further be filtered out by their flags, MAPQ qualities, or
custom filtering function. Using the CIGAR string, the aligned portion of the read
is traversed and the presence of each nucleotide base is tallied into respective arrays.
Additionally, the user can choose to filter the counted bases based on its base quality
score that is stored in the quality string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) â€“ the reference name (Default: None)</p></li>
<li><p><strong>reference</strong> (<em>str</em>) â€“ synonym for <cite>contig</cite> (Default: None)</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ 0-based inclusive start position (Default: None)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ 0-based exclusive start position (Default: None)</p></li>
<li><p><strong>end</strong> (<em>int</em>) â€“ Synonym for <cite>stop</cite> (Default: None)</p></li>
<li><p><strong>region</strong> (<em>str</em>) â€“ SAM-style region format.                 Example: â€˜chr1:10000-50000â€™ (Default: None)</p></li>
<li><p><strong>quality_threshold</strong> (<em>int</em>) â€“ MAPQ quality threshold (Default: 15)</p></li>
<li><p><strong>base_quality_threshold</strong> (<em>int</em>) â€“ base quality score threshold (Default: 0)</p></li>
<li><p><strong>read_callback</strong> (<em>str</em><em>|</em><em>function</em>) â€“ <p>select (or create) a filter of which                 reads to count. Built-in filters:</p>
<ul>
<li><dl class="simple">
<dt><cite>all</cite>: skips reads that contain the following flags:</dt><dd><ul>
<li><p>0x4 (4): read unmapped</p></li>
<li><p>0x100 (256): not primary alignment</p></li>
<li><p>0x200 (512): QC Fail</p></li>
<li><p>0x400 (1024): PCR or optical duplicate</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><cite>nofilter</cite>: uses all reads (Default)</p></li>
<li><p>The user can also supply a custom function that                     returns boolean objects for each read</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Four arrays in the order of <strong>A</strong>, <strong>C</strong>, <strong>G</strong>, <strong>T</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">array.array</span></code>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> â€“ if genomic coordinates are out of range or invalid, random access is disabled, or nucleotide base is unrecognized</p></li>
<li><p><strong>RuntimeError</strong> â€“ if <cite>read_callback</cite> is not properly set</p></li>
<li><p><strong>KeyError</strong> â€“ Reference is not found in header</p></li>
<li><p><strong>AssertionError</strong> â€“ if genomic region is malformed</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>SAM region formatted strings take on the following form:
â€˜chr1:100-200â€™</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>AlignmentFile.count_coverage(contig=â€™chr1â€™, start=1, stop= 1000)
AlignmentFile.count_coverage(â€˜chr1â€™, 1, 1000)
AlignmentFile.count_coverage(â€˜chr1:1-1000â€™)
AlignmentFile.count_coverage(â€˜chr1â€™, 1)
AlignmentFile.count_coverage(â€˜chr1â€™)</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">arr</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">count_coverage</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">):</span> 
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;array(&#39;</span><span class="si">{}</span><span class="s2">&#39;, </span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">typecode</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">arr</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))))</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 0, 0, 0, 0, 0])</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 15, 0, 14, 0, 0])</span>
<span class="go">array(&#39;L&#39;, [1, 1, 2, 2, ..., 0, 0, 0, 0, 14])</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 0, 15, 0, 14, 0])</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">arr</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">count_coverage</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="n">quality_threshold</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">base_quality_threshold</span><span class="o">=</span><span class="mi">25</span><span class="p">):</span> 
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;array(&#39;</span><span class="si">{}</span><span class="s2">&#39;, </span><span class="si">{}</span><span class="s2">)&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">typecode</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="n">arr</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))))</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 0, 0, 0, 0, 0])</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 13, 0, 14, 0, 0])</span>
<span class="go">array(&#39;L&#39;, [1, 1, 2, 2, ..., 0, 0, 0, 0, 13])</span>
<span class="go">array(&#39;L&#39;, [0, 0, 0, 0, ..., 0, 14, 0, 13, 0])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.get_index_stats">
<span class="sig-name descname"><span class="pre">get_index_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.get_index_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.get_index_stats" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Inspects the index file (BAI) for alignment statistics.</p>
<p>Every BAM index file contains metrics regarding the alignment
process for the given BAM file. The stored data are the number
of mapped and unmapped reads for a given reference. Unmapped reads
are paired end reads where only one part is mapped. Additionally,
index files also contain the number of unplaced unmapped reads. This
is stored within the <cite>nocoordinate</cite> instance attribute (if present).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of tuples for each reference in the order seen in the header. Each tuple contains the number of mapped reads, unmapped reads, and the sum of both.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>idx_stats (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">tuple</span></code>)</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>AssertionError</strong> â€“ if the index file is not available</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">get_index_stats</span><span class="p">()</span>
<span class="go">[(1446, 18, 1464), (1789, 17, 1806)]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam_no_bai</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">,</span> <span class="n">index_filename</span><span class="o">=</span><span class="s1">&#39;not_a_file.bai&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam_no_bai</span><span class="o">.</span><span class="n">get_index_stats</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">AssertionError</span>: <span class="n">No index available</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.is_valid_tid">
<span class="sig-name descname"><span class="pre">is_valid_tid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tid</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.is_valid_tid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.is_valid_tid" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return <cite>True</cite> if TID/RefID is valid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><cite>True</cite> if TID/refID is valid, else <cite>False</cite></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">is_valid_tid</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">is_valid_tid</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span> <span class="c1"># because there are only 2 in this file</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.get_reference_name">
<span class="sig-name descname"><span class="pre">get_reference_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tid</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.get_reference_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.get_reference_name" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Convert TID/refID to reference name.</p>
<p>The TID/refID is the position a reference sequence is seen
within the header file of the BAM file. The references are
sorted by ASCII order. Therefore, for a <strong>Homo sapien</strong> aligned
to GRCh38, â€˜chr10â€™ comes before â€˜chr1â€™ in the header. Therefore,
â€˜chr10â€™ would have the TID/refID of 0, not â€˜chr1â€™.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tid</strong> (<em>int</em>) â€“ TID/refID of desired reference/contig</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>String representation of chromosome if valid, else None</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>KeyError</strong> â€“ if TID/refID is not valid</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">get_reference_name</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> 
<span class="go">&#39;chr1&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">get_reference_name</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;10 is not a valid TID/refID for this file.&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.get_tid">
<span class="sig-name descname"><span class="pre">get_tid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">reference</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.get_tid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.get_tid" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Convert reference/contig name to refID/TID.</p>
<p>The TID/refID is the position a reference sequence is seen
within the header file of the BAM file. The references are
sorted by ASCII order. Therefore, for a <strong>Homo sapien</strong> aligned
to GRCh38, â€˜chr10â€™ comes before â€˜chr1â€™ in the header. Therefore,
â€˜chr10â€™ would have the TID/refID of 0, not â€˜chr1â€™.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>reference</strong> (<em>str</em>) â€“ reference/contig name</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the TID/refID of desired reference/contig</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> â€“ if reference name not found file header</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">get_tid</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">get_tid</span><span class="p">(</span><span class="s1">&#39;chr10&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">KeyError</span>: <span class="n">&#39;chr10 was not found in the file header&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.mate">
<span class="sig-name descname"><span class="pre">mate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">AlignedSegment</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.mate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.mate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Gets the mate to a given AlignedSegment.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Slow, when compared to the C-API. Not meant for high-throughput analysis.</p>
</div>
<p>Does not advance current iterator position.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>AlignedSegment</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.AlignedSegment</span></code>) â€“ a bamnostic AlignedSegment read with a mate</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if read has a valid mate, else None</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.AlignedSegment</span></code>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if AlignedSegment is unpaired</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.head">
<span class="sig-name descname"><span class="pre">head</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple_iterators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.head"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.head" title="Permalink to this definition">Â¶</a></dt>
<dd><p>List out the first <strong>n</strong> reads of the file.</p>
<p>This method is primarily used when doing an initial exploration
of the data. Whether or not <cite>multiple_iterators</cite> is used, cursor
position within the file will not change.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Using <cite>multiple_interators</cite> opens a new file object of the
same file currently in use and, thus, impacts the memory
footprint of your analysis.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> (<em>int</em>) â€“ number of aligned reads to print (default: 5)</p></li>
<li><p><strong>mutliple_iterators</strong> (<em>bool</em>) â€“ Whether to use current file object or create a new one (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of <strong>n</strong> reads from the front of the BAM file</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>head_reads (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignedSegment</span></code>)</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span> <span class="o">=</span> <span class="n">bamnostic</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">bamnostic</span><span class="o">.</span><span class="n">example_bam</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">multiple_iterators</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> 
<span class="go">EAS56_57:6:190:289:82       ...     MF:C:192</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bam</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">n</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> 
<span class="go">EAS56_57:6:190:289:82       ...     UQ:C:0</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.next">
<span class="sig-name descname"><span class="pre">next</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.next"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.next" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return the next line.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamReader.seekable">
<span class="sig-name descname"><span class="pre">seekable</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamReader.seekable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamReader.seekable" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return True indicating the BGZF supports random access.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Modified from original Bio.BgzfReader: checks to see if BAM
file has associated index file (BAI)</p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bam.BamWriter">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bam.</span></span><span class="sig-name descname"><span class="pre">BamWriter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath_or_object</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'xb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compresslevel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">b''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamWriter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamWriter" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#bamnostic.bgzf.BgzfWriter" title="bamnostic.bgzf.BgzfWriter"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.bgzf.BgzfWriter</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamWriter.write_header">
<span class="sig-name descname"><span class="pre">write_header</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">b''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_lengths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamWriter.write_header"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamWriter.write_header" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Modified code from Peter Cockâ€™s [BamWriter](<a class="reference external" href="https://github.com/peterjc/biopython/blob/SamBam2015/Bio/Sequencing/SamBam/__init__.py#L1714">https://github.com/peterjc/biopython/blob/SamBam2015/Bio/Sequencing/SamBam/__init__.py#L1714</a>)</p>
<p>Writes the header into a BAM file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_header</strong> (<em>filepath_or_object</em>) â€“ copy the header from another BAM file</p></li>
<li><p><strong>header</strong> (<em>bytes</em>) â€“ SAM-style header</p></li>
<li><p><strong>reference_names</strong> (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code>) â€“ list of all reference names used in file</p></li>
<li><p><strong>reference_lengths</strong> (<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">int</span></code>) â€“ list of all reference lengths used in file</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bam.BamWriter.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bam.html#BamWriter.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bam.BamWriter.write" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Write method for the class.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bam.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bam.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bam.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bam.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bam.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bam.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

</div>
<div class="section" id="module-bamnostic.bai">
<span id="bam-index-bai-file-handler-bamnostic-bai"></span><h2>BAM Index (BAI) file handler (<code class="docutils literal notranslate"><span class="pre">bamnostic.bai</span></code>)<a class="headerlink" href="#module-bamnostic.bai" title="Permalink to this headline">Â¶</a></h2>
<p>BAI file parser</p>
<p>Copyright Â© 2018, Marcus D. Sherman</p>
<p>This code is part of the bamnostic distribution and governed by its
license.  Please see the LICENSE file that should have been included
as part of this package.</p>
<p>The Binary Alignment Map (BAM) format (defined at <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a>)
allows for indexing. When the user invokes a tool to index a given BAM file, a BAM index (BAI)
file is created. Generally speaking, a BAI contains the all the virtual offsets of clusters of
reads that are associated with specific subsections of the BAM file. When the BAI is produced,
random access to the BAM is available.</p>
<p>This script is for parsing the binary encoded BAI file, inherently making it human readable.
Furthermore, it allows subsections of the BAI to be loaded into memory, reducing the memory
footprint and speeding up queries within a small number of references. Lastly, by parsing it
as such, random access queries directly into the associated BAM file is available to other
tools within bamnostic</p>
<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_chunk">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_chunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_chunk" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_intervals">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_intervals" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_bid_nchunk">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_bid_nchunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_bid_nchunk" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_unmapped">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_unmapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bai.RefIdx">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">RefIdx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.RefIdx" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.bai.RefIdx.end_offset">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">end_offset</span></span><a class="headerlink" href="#bamnostic.bai.RefIdx.end_offset" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.bai.RefIdx.n_bins">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">n_bins</span></span><a class="headerlink" href="#bamnostic.bai.RefIdx.n_bins" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bamnostic.bai.RefIdx.start_offset">
<em class="property"><span class="pre">property</span> </em><span class="sig-name descname"><span class="pre">start_offset</span></span><a class="headerlink" href="#bamnostic.bai.RefIdx.start_offset" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.reg2bin">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">reg2bin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rbeg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rend</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#reg2bin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.reg2bin" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds the largest superset bin of region. Numeric values taken from hts-specs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rbeg</strong> (<em>int</em>) â€“ inclusive beginning position of region</p></li>
<li><p><strong>rend</strong> (<em>int</em>) â€“ exclusive end position of region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>distinct bin ID for largest superset bin of region</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.reg2bins">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">reg2bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rbeg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rend</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#reg2bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.reg2bins" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generates bin ids which overlap the specified region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rbeg</strong> (<em>int</em>) â€“ inclusive beginning position of region</p></li>
<li><p><strong>rend</strong> (<em>int</em>) â€“ exclusive end position of region</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>(int)</em> â€“ bin IDs for overlapping bins of region</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if the range is malformed or invalid</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.bai.Bai">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">Bai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>This class defines the bam index file object and its interface.</p>
<p>The purpose of this class is the binary parsing of the bam index file (BAI) associated with
a given bam file. When queried, the Bai object identifies the bins of data that overlap the
requested region and directs which parts of the bam file contain it.</p>
<p>Virtual offsets are processed using the following method:
Beginning of compressed block = coffset = virtual offset &gt;&gt; 16
Position within uncompressed block = uoffset = virtual offset ^ (coffset &lt;&lt; 16)</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai._io">
<span class="sig-name descname"><span class="pre">_io</span></span><a class="headerlink" href="#bamnostic.bai.Bai._io" title="Permalink to this definition">Â¶</a></dt>
<dd><p>opened BAI file object</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>fileObject</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai._LINEAR_INDEX_WINDOW">
<span class="sig-name descname"><span class="pre">_LINEAR_INDEX_WINDOW</span></span><a class="headerlink" href="#bamnostic.bai.Bai._LINEAR_INDEX_WINDOW" title="Permalink to this definition">Â¶</a></dt>
<dd><p>constant of the linear interval window size</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai._UNMAP_BIN">
<span class="sig-name descname"><span class="pre">_UNMAP_BIN</span></span><a class="headerlink" href="#bamnostic.bai.Bai._UNMAP_BIN" title="Permalink to this definition">Â¶</a></dt>
<dd><p>constant for bin ID of unmapped read stats</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.magic">
<span class="sig-name descname"><span class="pre">magic</span></span><a class="headerlink" href="#bamnostic.bai.Bai.magic" title="Permalink to this definition">Â¶</a></dt>
<dd><p>first 4 bytes of file. Must be equal to bâ€™BAIâ€™</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bytes</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.n_refs">
<span class="sig-name descname"><span class="pre">n_refs</span></span><a class="headerlink" href="#bamnostic.bai.Bai.n_refs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of references in BAI</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.unmapped">
<span class="sig-name descname"><span class="pre">unmapped</span></span><a class="headerlink" href="#bamnostic.bai.Bai.unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of the unmapped read stats by each reference</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.current_ref">
<span class="sig-name descname"><span class="pre">current_ref</span></span><a class="headerlink" href="#bamnostic.bai.Bai.current_ref" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of the current reference loaded into memory.
It contains the a dictionary of bin IDs and their respective
chunks, and a list of linear intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>None|dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.ref_indices">
<span class="sig-name descname"><span class="pre">ref_indices</span></span><a class="headerlink" href="#bamnostic.bai.Bai.ref_indices" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of reference ids and their start/stop offsets within the BAI file</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.n_no_coord">
<span class="sig-name descname"><span class="pre">n_no_coord</span></span><a class="headerlink" href="#bamnostic.bai.Bai.n_no_coord" title="Permalink to this definition">Â¶</a></dt>
<dd><p>if present in BAI, is the number of reads that have no coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>None|int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.bai.Bai._last_pos">
<span class="sig-name descname"><span class="pre">_last_pos</span></span><a class="headerlink" href="#bamnostic.bai.Bai._last_pos" title="Permalink to this definition">Â¶</a></dt>
<dd><p>used for indexing, the byte position of the file head.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.get_chunks">
<span class="sig-name descname"><span class="pre">get_chunks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_chunks</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.get_chunks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.get_chunks" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple generator for unpacking chunk data</p>
<p>Chunks are defined as groupings of reads within a BAM file
that share the same bin. A <cite>Chunk</cite> object in the context of
this function is a <cite>namedtuple</cite> that contains the virtual offsets
for the beginning and end of each of these chunks.</p>
<p>Note: a special case of a chunk is in any Bin labeled as 37450.
These bins always contain 2 chunks that provide the statistics
of the number of reads that are unmapped to that reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_chunks</strong> (<em>int</em>) â€“ number of chunks to be unpacked from stream</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>a list of Chunk objects with the attributes of</dt><dd><p>chunks[i] are .voffset_beg and voffset_end</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>chunks (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.get_ints">
<span class="sig-name descname"><span class="pre">get_ints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_int</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.get_ints"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.get_ints" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Unpacks <cite>n_int</cite> number of interval virtual offsets from stream</p>
<p>A linear interval is defined as a 16384 bp window along a given reference. The
value stored in the BAI is the virtual offset of the first read within
that given interval. This virtual offset is the byte offset (coffset) of the start of the
BGZF block that contains the beginning of the read and the byte offset (uoffset)
within the uncompressed data of the residing BGZF block to that first read.</p>
<p>Note: a caveat to using linear interval with long reads: A long read can
span multiple linear intervals. As such, the current encoding could potentially shift
the expected region of interest to the left more than expected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_int</strong> (<em>int</em>) â€“ number of intervals to unpack</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of virtual offsets for <cite>n_int</cite> number of linear intervals</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>intervals (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.get_bins">
<span class="sig-name descname"><span class="pre">get_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.get_bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.get_bins" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple function that iteratively unpacks the data of a number (<cite>n_bin</cite>)
of bins.</p>
<p>As the function discovers the number of chunks needed for a given
bin, it deletages work to <cite>self.get_chunks(n_chunks)</cite>. A bin is
comprised of 2 parts: 1) the distinct bin ID (within a given reference). If
no reads are associated with that bin, it is left out of the indexing process,
and therefore not represented in the BAI file. Furthermore, while each bin
has a bin ID, the bin IDs are only distinct within a given reference. This means
that 2 or more references can have the same bin IDs. These bin IDs are also
not in any order as they are essentially a hash dump. Lastly, the only reserved
bin ID is 37450. This bin relates to 2 chunks that contain the number of
unmapped and mapped reads for a given reference. 2) the chunk(s) of reads that
are assigned to a given bin.</p>
<p>As a secondary feature, this function will also quickly seek over regions for
the purposes of documenting the start and stop byte offsets of a given reference
block within the file. This is invoked by setting <cite>idx=True</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_int</strong> (<em>int</em>) â€“ number of bins to be unpacked from stream</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>None if just indexing the index file or a dictionary</dt><dd><p>of <cite>bin_id: chunks</cite> pairs</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bins (None | dict)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if bin 37450 does not contain 2 chunks exactly</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.get_ref">
<span class="sig-name descname"><span class="pre">get_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.get_ref"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.get_ref" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Interatively unpacks all the bins, linear intervals, and chunks for a given reference</p>
<p>A reference is comprised of 2 things: 1) a series of bins that reference chunks of aligned
reads that are grouped within that bin. 2) a series of virtual offsets of the first read of a
16384 bp window along the given reference.</p>
<p>This function also serves to â€œindexâ€ the BAI file such that, if it is invoked by setting
<cite>ids=True</cite>, will do a single pass through the BAI file and saving the start and stop
offsets of each of the references. This is used for minimizing the memory footprint of
storing the BAI in memory. When queried against, the appropriate reference block will be
loaded. Because of this constant loading, <cite>functools.lru_cache</cite> was applied to cache recently
used reference blocks to speed up computation. It is assumed that when querying is done, most
users are looking and just a few references at a time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_id</strong> (<em>None</em><em>|</em><em>int</em>) â€“ used for random access or indexing the BAI</p></li>
<li><p><strong>idx</strong> (<em>bool</em>) â€“ Flag for setting whether or not to run an index of the BAI</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>namedtuple</cite> containing the byte offsets of the reference start, stop, and number of bins
or
Ref: <cite>namedtuple</cite> containing a dictionary of bins and list of linear intervals</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#bamnostic.bai.RefIdx" title="bamnostic.bai.RefIdx">RefIdx</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if, when random access is used, the current reference offset
    does not match indexed reference offset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.query">
<span class="sig-name descname"><span class="pre">query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.query" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Main query function for determining seek offset to BAM section that
AlignedRead objects from specified region start</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>int</em>) â€“ which reference/chromosome TID</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ left most bp position of region (zero-based)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ right most bp position of region (zero-based)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>the voffset_beg of the first chunk given the chunkâ€™s voffset_end</dt><dd><p>is greater than the voffset of the linear index that overlaps
the region of interestâ€™s start offset</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.seek">
<span class="sig-name descname"><span class="pre">seek</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">whence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.seek"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.seek" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple seek function for binary files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>offset</strong> (<em>None</em><em>|</em><em>int</em>) â€“ byte offset from whence to move the file head to.</p></li>
<li><p><strong>whence</strong> (<em>int</em>) â€“ 0 := from start of file, 1:= from current position, 2:= from end of file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>new byte position of file head</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if the offset is not an integer or is not provided</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.read" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple read function for binary files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>size</strong> (<em>int</em>) â€“ number of bytes to read in (default: -1 â€“whole file)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the number of bytes read from file</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(bytes)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.bai.Bai.tell">
<span class="sig-name descname"><span class="pre">tell</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/bai.html#Bai.tell"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.bai.Bai.tell" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple tell function for reporting byte position of file head</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>byte position of file head</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(int)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.bai.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.bai.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.bai.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

</div>
<div class="section" id="module-bamnostic.csi">
<span id="coordinate-sorted-index-csi-file-handler-bamnostic-csi"></span><h2>Coordinate-Sorted Index (CSI) file handler (<code class="docutils literal notranslate"><span class="pre">bamnostic.csi</span></code>)<a class="headerlink" href="#module-bamnostic.csi" title="Permalink to this headline">Â¶</a></h2>
<p>CSI file parser</p>
<p>Copyright Â© 2018, Marcus D. Sherman</p>
<p>This code is part of the bamnostic distribution and governed by its
license.  Please see the LICENSE file that should have been included
as part of this package.</p>
<p>The Binary Alignment Map (CSI) format (defined at <a class="reference external" href="http://samtools.github.io/hts-specs/CSIv1.pdf">http://samtools.github.io/hts-specs/CSIv1.pdf</a>)
allows for indexing. When the user invokes a tool to index a given BAM file, a CSI index file
can be created. Generally speaking, a CSI contains the all the virtual offsets of clusters of
reads that are associated with specific subsections of the BAM file. When the CSI is produced,
random access to the BAM is available.</p>
<p>This script is for parsing the binary encoded CSI file, inherently making it human readable.
Furthermore, it allows subsections of the CSI to be loaded into memory, reducing the memory
footprint and speeding up queries within a small number of references. Lastly, by parsing it
as such, random access queries directly into the associated BAM file is available to other
tools within bamnostic</p>
<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.unpack_chunk">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">unpack_chunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.csi.unpack_chunk" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.unpack_bid_loffset_nchunk">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">unpack_bid_loffset_nchunk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.csi.unpack_bid_loffset_nchunk" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.unpack_unmapped">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">unpack_unmapped</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.csi.unpack_unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.reg2bin_csi">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">reg2bin_csi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rbeg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rend</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#reg2bin_csi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.reg2bin_csi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Finds the largest superset bin of region. Numeric values taken from hts-specs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rbeg</strong> (<em>int</em>) â€“ inclusive beginning position of region</p></li>
<li><p><strong>rend</strong> (<em>int</em>) â€“ exclusive end position of region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>distinct bin ID for largest superset bin of region</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.reg2bins_csi">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">reg2bins_csi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rbeg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rend</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">14</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#reg2bins_csi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.reg2bins_csi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Generates bin ids which overlap the specified region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rbeg</strong> (<em>int</em>) â€“ inclusive beginning position of region</p></li>
<li><p><strong>rend</strong> (<em>int</em>) â€“ exclusive end position of region</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><em>(int)</em> â€“ bin IDs for overlapping bins of region</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if the range is malformed or invalid</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.csi.Csi">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">Csi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#Csi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.Csi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#bamnostic.bai.Bai" title="bamnostic.bai.Bai"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.bai.Bai</span></code></a></p>
<p>This class defines the bam CSI index file object and its interface.</p>
<p>The purpose of this class is the binary parsing of the bam index file (CSI) associated with
a given bam file. When queried, the Csi object identifies the bins of data that overlap the
requested region and directs which parts of the bam file contain it.</p>
<p>Virtual offsets are processed using the following method:
Beginning of compressed block = coffset = virtual offset &gt;&gt; 16
Position within uncompressed block = uoffset = virtual offset ^ (coffset &lt;&lt; 16)</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi._io">
<span class="sig-name descname"><span class="pre">_io</span></span><a class="headerlink" href="#bamnostic.csi.Csi._io" title="Permalink to this definition">Â¶</a></dt>
<dd><p>opened CSI file object</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>fileObject</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi._UNMAP_BIN">
<span class="sig-name descname"><span class="pre">_UNMAP_BIN</span></span><a class="headerlink" href="#bamnostic.csi.Csi._UNMAP_BIN" title="Permalink to this definition">Â¶</a></dt>
<dd><p>constant for bin ID of unmapped read stats</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.magic">
<span class="sig-name descname"><span class="pre">magic</span></span><a class="headerlink" href="#bamnostic.csi.Csi.magic" title="Permalink to this definition">Â¶</a></dt>
<dd><p>first 4 bytes of file. Must be equal to bâ€™CSIâ€™</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bytes</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.n_refs">
<span class="sig-name descname"><span class="pre">n_refs</span></span><a class="headerlink" href="#bamnostic.csi.Csi.n_refs" title="Permalink to this definition">Â¶</a></dt>
<dd><p>number of references in CSI</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.unmapped">
<span class="sig-name descname"><span class="pre">unmapped</span></span><a class="headerlink" href="#bamnostic.csi.Csi.unmapped" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of the unmapped read stats by each reference</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.current_ref">
<span class="sig-name descname"><span class="pre">current_ref</span></span><a class="headerlink" href="#bamnostic.csi.Csi.current_ref" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of the current reference loaded into memory.
It contains the a dictionary of bin IDs and their respective
chunks, and a list of linear intervals.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>None|dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.ref_indices">
<span class="sig-name descname"><span class="pre">ref_indices</span></span><a class="headerlink" href="#bamnostic.csi.Csi.ref_indices" title="Permalink to this definition">Â¶</a></dt>
<dd><p>dictionary of reference ids and their start/stop offsets within theCSI file</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.n_no_coord">
<span class="sig-name descname"><span class="pre">n_no_coord</span></span><a class="headerlink" href="#bamnostic.csi.Csi.n_no_coord" title="Permalink to this definition">Â¶</a></dt>
<dd><p>if present inCSI, is the number of reads that have no coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>None|int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bamnostic.csi.Csi._last_pos">
<span class="sig-name descname"><span class="pre">_last_pos</span></span><a class="headerlink" href="#bamnostic.csi.Csi._last_pos" title="Permalink to this definition">Â¶</a></dt>
<dd><p>used for indexing, the byte position of the file head.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.get_chunks">
<span class="sig-name descname"><span class="pre">get_chunks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_chunks</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#Csi.get_chunks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.Csi.get_chunks" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple generator for unpacking chunk data</p>
<p>Chunks are defined as groupings of reads within a BAM file
that share the same bin. A <cite>Chunk</cite> object in the context of
this function is a <cite>namedtuple</cite> that contains the virtual offsets
for the beginning and end of each of these chunks.</p>
<p>Note: a special case of a chunk is in any Bin labeled as 37450.
These bins always contain 2 chunks that provide the statistics
of the number of reads that are unmapped to that reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_chunks</strong> (<em>int</em>) â€“ number of chunks to be unpacked from stream</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>a list of Chunk objects with the attributes of</dt><dd><p>chunks[i] are .voffset_beg and voffset_end</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>chunks (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.get_bins">
<span class="sig-name descname"><span class="pre">get_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#Csi.get_bins"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.Csi.get_bins" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple function that iteratively unpacks the data of a number (<cite>n_bin</cite>)
of bins.</p>
<p>As the function discovers the number of chunks needed for a given
bin, it deletages work to <cite>self.get_chunks(n_chunks)</cite>. A bin is
comprised of 2 parts: 1) the distinct bin ID (within a given reference). If
no reads are associated with that bin, it is left out of the indexing process,
and therefore not represented in theCSI file. Furthermore, while each bin
has a bin ID, the bin IDs are only distinct within a given reference. This means
that 2 or more references can have the same bin IDs. These bin IDs are also
not in any order as they are essentially a hash dump. Lastly, the only reserved
bin ID is 37450. This bin relates to 2 chunks that contain the number of
unmapped and mapped reads for a given reference. 2) the chunk(s) of reads that
are assigned to a given bin.</p>
<p>As a secondary feature, this function will also quickly seek over regions for
the purposes of documenting the start and stop byte offsets of a given reference
block within the file. This is invoked by setting <cite>idx=True</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_int</strong> (<em>int</em>) â€“ number of bins to be unpacked from stream</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>None if just indexing the index file or a dictionary</dt><dd><p>of <cite>bin_id: chunks</cite> pairs</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bins (None | dict)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if bin 37450 does not contain 2 chunks exactly</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.get_ref">
<span class="sig-name descname"><span class="pre">get_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#Csi.get_ref"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.Csi.get_ref" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Iteratively unpacks all the bins, linear intervals, and chunks for a given reference</p>
<p>A reference is comprised of 2 things: 1) a series of bins that reference chunks of aligned
reads that are grouped within that bin. 2) a series of virtual offsets of the first read of a
16384 bp window along the given reference.</p>
<p>This function also serves to â€œindexâ€ theCSI file such that, if it is invoked by setting
<cite>ids=True</cite>, will do a single pass through theCSI file and saving the start and stop
offsets of each of the references. This is used for minimizing the memory footprint of
storing theCSI in memory. When queried against, the appropriate reference block will be
loaded. Because of this constant loading, <cite>functools.lru_cache</cite> was applied to cache recently
used reference blocks to speed up computation. It is assumed that when querying is done, most
users are looking and just a few references at a time.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_id</strong> (<em>None</em><em>|</em><em>int</em>) â€“ used for random access or indexing theCSI</p></li>
<li><p><strong>idx</strong> (<em>bool</em>) â€“ Flag for setting whether or not to run an index of theCSI</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>namedtuple</cite> containing the byte offsets of the reference start, stop, and number of bins
o
Ref: <cite>namedtuple</cite> containing a dictionary of bins and list of linear intervals</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#bamnostic.bai.RefIdx" title="bamnostic.bai.RefIdx">RefIdx</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>AssertionError</strong><strong> (</strong><strong>Exception</strong><strong>)</strong> â€“ if, when random access is used, the current reference offset
    does not match indexed reference offset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.csi.Csi.query">
<span class="sig-name descname"><span class="pre">query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-</span> <span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/csi.html#Csi.query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.csi.Csi.query" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Main query function for determining seek offset to BAM section that
AlignedRead objects from specified region start</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>int</em>) â€“ which reference/chromosome TID</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ left most bp position of region (zero-based)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ right most bp position of region (zero-based)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>the voffset_beg of the first chunk given the chunkâ€™s voffset_end</dt><dd><p>is greater than the voffset of the linear index that overlaps
the region of interestâ€™s start offset</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(int)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.csi.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.csi.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.csi.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.csi.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

</div>
<div class="section" id="module-bamnostic.utils">
<span id="bamnostic-utilities-bamnostic-utils"></span><h2>BAMnostic Utilities (<code class="docutils literal notranslate"><span class="pre">bamnostic.utils</span></code>)<a class="headerlink" href="#module-bamnostic.utils" title="Permalink to this headline">Â¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.format_warnings">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">format_warnings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">message</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">category</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lineno</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#format_warnings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.format_warnings" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Sets STDOUT warnings</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>message</strong> â€“ the unformatted warning message being reported</p></li>
<li><p><strong>category</strong> (<em>str</em>) â€“ the level of warning (handled by <cite>warnings</cite> module)</p></li>
<li><p><strong>filename</strong> (<em>str</em>) â€“ filename for logging purposes (defaults to STDOUT)</p></li>
<li><p><strong>lineno</strong> (<em>int</em>) â€“ where the error occurred.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>formatted warning string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.unpack_int32">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">unpack_int32</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.utils.unpack_int32" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.unpack_int32L">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">unpack_int32L</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="pre">buffer</span></em>, <em class="sig-param"><span class="pre">/</span></em><span class="sig-paren">)</span><a class="headerlink" href="#bamnostic.utils.unpack_int32L" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Return a tuple containing unpacked values.</p>
<p>Unpack according to the format string Struct.format. The bufferâ€™s size
in bytes must be Struct.size.</p>
<p>See help(struct) for more on format strings.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.utils.Roi">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">Roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#Roi"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.Roi" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Small __slots__ class for region of interest parsing</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.flag_decode">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">flag_decode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flag_code</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#flag_decode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.flag_decode" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple read alignment flag decoder</p>
<p>Every read within a BAM file ought to have an associated flag code. Theses
flags are used for read filtering and QC. The flags are described below.
Additionally, they can be found <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">here</a></p>
<p>Any given readâ€™s flag is determined by the <em>or</em> (<cite>|</cite>) operand of all appropriate bit flags.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flag_code</strong> (<em>int</em>) â€“ either a standalone integer/bit flag or the read object itself</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of flag and flag description tuples.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">tuple</span></code>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if provided flag is not a valid entry</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>If a flag is 516 it is comprised of flag 4 and flag 512</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">flag_decode</span><span class="p">(</span><span class="mi">516</span><span class="p">)</span>
<span class="go">[(4, &#39;read unmapped&#39;), (512, &#39;QC fail&#39;)]</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 5%" />
<col style="width: 7%" />
<col style="width: 88%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Int</p></th>
<th class="head"><p>Bit</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>0x1</p></td>
<td><p>Template having multiple segments in sequencing</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>0x2</p></td>
<td><p>Each segment properly aligned according to the aligner</p></td>
</tr>
<tr class="row-even"><td><p>4</p></td>
<td><p>0x4</p></td>
<td><p>Segment unmapped</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p>0x8</p></td>
<td><p>Next segment in the template unmapped</p></td>
</tr>
<tr class="row-even"><td><p>16</p></td>
<td><p>0x10</p></td>
<td><p>SEQ being reverse complemented</p></td>
</tr>
<tr class="row-odd"><td><p>32</p></td>
<td><p>0x20</p></td>
<td><p>SEQ of the next segment in the template being reverse complemented</p></td>
</tr>
<tr class="row-even"><td><p>64</p></td>
<td><p>0x40</p></td>
<td><p>The first segment in the template</p></td>
</tr>
<tr class="row-odd"><td><p>128</p></td>
<td><p>0x80</p></td>
<td><p>The last segment in the template</p></td>
</tr>
<tr class="row-even"><td><p>256</p></td>
<td><p>0x100</p></td>
<td><p>Secondary alignment</p></td>
</tr>
<tr class="row-odd"><td><p>512</p></td>
<td><p>0x200</p></td>
<td><p>Not passing filters, such as platform/vendor quality controls</p></td>
</tr>
<tr class="row-even"><td><p>1024</p></td>
<td><p>0x400</p></td>
<td><p>PCR or optical duplicate</p></td>
</tr>
<tr class="row-odd"><td><p>2048</p></td>
<td><p>0x800</p></td>
<td><p>Supplementary alignment</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.yes_no">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">yes_no</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">message</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#yes_no"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.yes_no" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Simple prompt parser</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.parse_region">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">parse_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">until_eof</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#parse_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.parse_region" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Parses region information from all user set parameters.</p>
<p>The main goal of this function is to handle the many different ways a user
can put in genomic region data. One way is through keyword arguments. This
is the most straight forward. However, due to Pysamâ€™s API, there are multiple
synonyms for reference/contig or stop/end. Additionally, if the user knows
the refID/TID of the reference they are interested in, they can input that way
as well.</p>
<p>The second form a submission can take is through positional arguments. Just like
keyword, but ordered such that it makes up a genomic region of interest.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Positional arguments make utilizing the <cite>tid</cite> parameter difficult since it is
the 5th argument of the function signature.</p>
</div>
<p>The third form a submission can take is through using a SAM-formatted string. An
example SAM-formatted string looks like â€˜chr1:10-100â€™. As many users also copy &amp;
paste directly from tab-delimited files, such as BED files, a SAM-formatted string
can take the form of â€˜chr1\t10\t100â€™ where â€˜\tâ€™ indicates a tab space.</p>
<p>Lastly, the <cite>until_eof</cite> switch allows users to take all items from their desired
start position (be it the whole reference or a specific spot on the reference). Setting
this to <cite>True</cite> (default: <cite>False</cite>) will pull all reads to the end of the reference or file,
whichever is first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contig</strong> (<em>str</em>) â€“ name of reference/contig</p></li>
<li><p><strong>start</strong> (<em>int</em>) â€“ start position of region of interest (0-based)</p></li>
<li><p><strong>stop</strong> (<em>int</em>) â€“ stop position of region of interest (0-based)</p></li>
<li><p><strong>region</strong> (<em>str</em>) â€“ SAM region formatted string. Accepts tab-delimited values as well</p></li>
<li><p><strong>tid</strong> (<em>int</em>) â€“ the refID or target id of a reference/contig</p></li>
<li><p><strong>until_eof</strong> (<em>bool</em>) â€“ iterate until end of file (default: False)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>region of interest formatted as an object with named attributes.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>query (<a class="reference internal" href="#bamnostic.utils.Roi" title="bamnostic.utils.Roi"><code class="xref py py-class docutils literal notranslate"><span class="pre">bamnostic.utils.Roi</span></code></a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if two synonym keywords are set, but contradict each other</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Keyword-based</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">contig</span> <span class="o">=</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">contig</span><span class="p">:</span> <span class="n">chr1</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># Using `tid` instead of `contig`</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">tid</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">tid</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># Positional arguments</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">contig</span><span class="p">:</span> <span class="n">chr1</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># SAM-formatted string (keyword)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">region</span> <span class="o">=</span> <span class="s1">&#39;chr1:10-100&#39;</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">contig</span><span class="p">:</span> <span class="n">chr1</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># SAM-formatted string (positional)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="s1">&#39;chr1:10-100&#39;</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">contig</span><span class="p">:</span> <span class="n">chr1</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># Tab-delimited region string</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="s1">&#39;chr1</span><span class="se">\t</span><span class="s1">10</span><span class="se">\t</span><span class="s1">100&#39;</span><span class="p">)</span>
<span class="n">Roi</span><span class="p">(</span><span class="n">contig</span><span class="p">:</span> <span class="n">chr1</span><span class="p">,</span> <span class="n">start</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="n">stop</span><span class="p">:</span> <span class="mi">100</span><span class="p">)</span>

<span class="c1"># Contradictory synonyms</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">parse_region</span><span class="p">(</span><span class="n">reference</span><span class="o">=</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="n">contig</span><span class="o">=</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">stop</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
<span class="n">Traceback</span> <span class="p">(</span><span class="n">most</span> <span class="n">recent</span> <span class="n">call</span> <span class="n">last</span><span class="p">):</span>
<span class="o">...</span>
<span class="ne">ValueError</span><span class="p">:</span> <span class="n">either</span> <span class="n">contig</span> <span class="ow">or</span> <span class="n">reference</span> <span class="n">must</span> <span class="n">be</span> <span class="nb">set</span><span class="p">,</span> <span class="ow">not</span> <span class="n">both</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.unpack">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">unpack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fmt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_io</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_array</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#unpack"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.unpack" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Utility function for unpacking binary data from file object or byte
stream.</p>
<p>The only difference between this method and <cite>struct.unpack</cite> is that
<cite>unpack</cite> dynamically determines the size needed to read in based on
the format string. Additionally, it can process a file object or byte
stream and implement a read or slice (respectively). Mainly, this is a
quality of life function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fmt</strong> (<em>str</em>) â€“ the string format of the binary data to be unpacked</p></li>
<li><p><strong>_io</strong> â€“ built-in binary format reader (default: io.BufferedRandom)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>unpacked contents from _io based on fmt string</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.make_virtual_offset">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">make_virtual_offset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">block_start_offset</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">within_block_offset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#make_virtual_offset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.make_virtual_offset" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Compute a BGZF virtual offset from block start and within block offsets.</p>
<p>The BAM indexing scheme records read positions using a 64 bit
â€˜virtual offsetâ€™, comprising in C terms:</p>
<p>block_start_offset &lt;&lt; 16 | within_block_offset</p>
<p>Here block_start_offset is the file offset of the BGZF block
start (unsigned integer using up to 64-16 = 48 bits), and
within_block_offset within the (decompressed) block (unsigned
16 bit integer).</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">0</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">16</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
<span class="go">65535</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_virtual_offset</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">16</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">Require 0 &lt;= within_block_offset &lt; 2**16, got 65536</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.split_virtual_offset">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">split_virtual_offset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">virtual_offset</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#split_virtual_offset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.split_virtual_offset" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Divides a 64-bit BGZF virtual offset into block start &amp; within block offsets.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">==</span> <span class="n">split_virtual_offset</span><span class="p">(</span><span class="mi">6553600000</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="mi">100000</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span> <span class="o">==</span> <span class="n">split_virtual_offset</span><span class="p">(</span><span class="mi">6553600010</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bamnostic.utils.LruDict">
<em class="property"><span class="pre">class</span> </em><span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">LruDict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#LruDict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.LruDict" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">collections.OrderedDict</span></code></p>
<p>Simple least recently used (LRU) based dictionary that caches a given
number of items.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.utils.LruDict.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#LruDict.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.LruDict.get" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Basic getter that renews LRU status upon inspection</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> (<em>str</em>) â€“ immutable dictionary key</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.utils.LruDict.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">others</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#LruDict.update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.LruDict.update" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Same as a regular <cite>dict.update</cite>, however, since pypyâ€™s <cite>dict.update</cite>
doesnâ€™t go through <cite>dict.__setitem__</cite>, this is used to ensure it does</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>others</strong> (<em>iterable</em>) â€“ a dictionary or iterable containing key/value pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bamnostic.utils.LruDict.cull">
<span class="sig-name descname"><span class="pre">cull</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#LruDict.cull"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.LruDict.cull" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Main utility function for pruning the LruDict</p>
<p>If the length of the LruDict is more than <cite>max_cache</cite>, it removes the LRU item</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.parse_cigar">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">parse_cigar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cigar_str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#parse_cigar"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.parse_cigar" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Parses a CIGAR string and turns it into a list of tuples</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cigar_str</strong> (<em>str</em>) â€“ the CIGAR string as shown in SAM entry</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples of CIGAR operations (by id) and number of operations</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>cigar_array (list)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if CIGAR operation is invalid</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">parse_cigar</span><span class="p">(</span><span class="s1">&#39;3M1I3M1D5M&#39;</span><span class="p">)</span> 
<span class="go">[((&#39;BAM_CMATCH&#39;, 0), 3), ..., ((&#39;BAM_CMATCH&#39;, 0), 5)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.check_cigar_arg">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">check_cigar_arg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cigar</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#check_cigar_arg"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.check_cigar_arg" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Checks to make sure CIGAR arugment is valid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>argument</strong> (str or <code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code>) â€“ CIGAR string (pre-formatted or raw)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>CIGAR re-formatted as a list</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">list</span></code>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if CIGAR is not a string or pre-formatted list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.cigar_changes">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">cigar_changes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cigar</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#cigar_changes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.cigar_changes" title="Permalink to this definition">Â¶</a></dt>
<dd><dl class="simple">
<dt>Recreates the reference sequence to the extent that the CIGAR string can</dt><dd><p>represent.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) â€“ aligned segment sequence</p></li>
<li><p><strong>cigar</strong> (<em>list</em>) â€“ list of tuples of cigar operations (by id) and number of operations</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a version of the aligned segmentâ€™s reference             sequence given the changes reflected in the cigar string</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>cigar_formatted_ref (str)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if CIGAR operation is invalid</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cigar_changes</span><span class="p">(</span><span class="s1">&#39;ACTAGAATGGCT&#39;</span><span class="p">,</span> <span class="s1">&#39;3M1I3M1D5M&#39;</span><span class="p">)</span>
<span class="go">&#39;ACTGAATGGCT&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cigar_changes</span><span class="p">(</span><span class="s1">&#39;ACTAGAATGGCT&#39;</span><span class="p">,</span> <span class="s1">&#39;3V1I3M1D5M&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">Invalid CIGAR operation (V).</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.md_changes">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">md_changes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_tag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#md_changes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.md_changes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Recreates the reference sequence of a given alignment to the extent that the
MD tag can represent.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Used in conjunction with <cite>cigar_changes</cite> to recreate the
complete reference sequence</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) â€“ aligned segment sequence</p></li>
<li><p><strong>md_tag</strong> (<em>str</em>) â€“ MD tag for associated sequence</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a version of the aligned segmentâ€™s reference sequence given             the changes reflected in the MD tag</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ref_seq (str)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> â€“ if MD tag is None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">md_changes</span><span class="p">(</span><span class="s1">&#39;CTTATATTGGCCTT&#39;</span><span class="p">,</span> <span class="s1">&#39;3C4AT4&#39;</span><span class="p">)</span>
<span class="go">&#39;CTTCTATTATCCTT&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.ref_gen">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">ref_gen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cigar_string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_tag</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#ref_gen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.ref_gen" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Recreates the reference sequence associated with the given segment.</p>
<p>Uses the CIGAR string and MD tag to recreate the reference sequence associated
with the aligned segment. This is done without the need for looking up
the reference genome. Example reads, MD tags, and CIGAR strings taken from
<a class="reference external" href="https://davetang.org/muse/2011/01/28/perl-and-sam/">David Tangâ€™s Blog</a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generated reference sequence</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>(str)</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>KeyError</strong> â€“ if read does not contain MD tag</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p># Only mismatches
&gt;&gt;&gt; seq = â€˜CGATACGGGGACATCCGGCCTGCTCCTTCTCACATGâ€™
&gt;&gt;&gt; cigar = â€˜36Mâ€™
&gt;&gt;&gt; md = â€˜1A0C0C0C1T0C0T27â€™
&gt;&gt;&gt; ref_gen(seq, cigar, md)
â€˜CACCCCTCTGACATCCGGCCTGCTCCTTCTCACATGâ€™</p>
<p># Insertions and mismatches
&gt;&gt;&gt; seq = â€˜GAGACGGGGTGACATCCGGCCTGCTCCTTCTCACATâ€™
&gt;&gt;&gt; cigar = â€˜6M1I29Mâ€™
&gt;&gt;&gt; md = â€˜0C1C0C1C0T0C27â€™
&gt;&gt;&gt; ref_gen(seq, cigar, md)
â€˜CACCCCTCTGACATCCGGCCTGCTCCTTCTCACATâ€™</p>
<p># Deletion and mismatches
&gt;&gt;&gt; seq = â€˜AGTGATGGGGGGGTTCCAGGTGGAGACGAGGACTCCâ€™
&gt;&gt;&gt; cigar = â€˜9M9D27Mâ€™
&gt;&gt;&gt; md = â€˜2G0A5^ATGATGTCA27â€™
&gt;&gt;&gt; ref_gen(seq, cigar, md)
â€˜AGGAATGGGATGATGTCAGGGGTTCCAGGTGGAGACGAGGACTCCâ€™</p>
<p># Insertion, deletion, and mismatches
&gt;&gt;&gt; seq = â€˜AGTGATGGGAGGATGTCTCGTCTGTGAGTTACAGCAâ€™
&gt;&gt;&gt; cigar = â€˜2M1I7M6D26Mâ€™
&gt;&gt;&gt; md = â€˜3C3T1^GCTCAG26â€™
&gt;&gt;&gt; ref_gen(seq, cigar, md)
â€˜AGGCTGGTAGCTCAGGGATGTCTCGTCTGTGAGTTACAGCAâ€™</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bamnostic.utils.cigar_alignment">
<span class="sig-prename descclassname"><span class="pre">bamnostic.utils.</span></span><span class="sig-name descname"><span class="pre">cigar_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cigar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualities</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_qual_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bamnostic/utils.html#cigar_alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bamnostic.utils.cigar_alignment" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Use the CIGAR to filter out all unaligned data bases</p>
<p>Any clipping results in the removal of those bases. If an insertion is seen in
the CIGAR, those bases are removed from the sequence. If a deletion is seen in
the CIGAR, those bases are padded with a period (â€˜.â€™) symbol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) â€“ string sequence of the aligned segment.</p></li>
<li><p><strong>cigar</strong> (<em>str</em>) â€“ the cigar string or <cite>cigartuple</cite> of the aligned segment.</p></li>
<li><p><strong>start_pos</strong> (<em>int</em>) â€“ the first aligned position of the read</p></li>
<li><p><strong>qualities</strong> (<code class="xref py py-obj docutils literal notranslate"><span class="pre">array.array</span></code>) â€“ base quality array from read</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>(<code class="xref py py-obj docutils literal notranslate"><span class="pre">tuple</span></code> of <code class="xref py py-obj docutils literal notranslate"><span class="pre">str</span></code> and <code class="xref py py-obj docutils literal notranslate"><span class="pre">int</span></code>) â€“ nucleotide base and index position of that base relative to reference</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;AGTGATGGGAGGATGTCTCGTCTGTGAGTTACAGCA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cigar</span> <span class="o">=</span> <span class="s1">&#39;2M1I7M6D26M&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_position</span> <span class="o">=</span> <span class="mi">95</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c_a</span> <span class="o">=</span> <span class="n">cigar_alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">cigar</span><span class="p">,</span> <span class="n">start_position</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">next</span><span class="p">(</span><span class="n">c_a</span><span class="p">)</span>
<span class="go">(&#39;A&#39;, 95)</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="quickstart.html" class="btn btn-neutral float-left" title="Quickstart" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="support.html" class="btn btn-neutral float-right" title="Citation and Support" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Marcus D. Sherman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>